% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/getFragment.R
\name{getFragment}
\alias{getFragment}
\title{Generating spliced fragments of the input alignments}
\usage{
getFragment(alignment)
}
\arguments{
\item{alignment}{the alignments to get spliced fragments}
}
\value{
a list of two data frame for fragments come from positive and negative reads. Each data frame has information
about the start/end position of the fragments and the group which allows the refer which read that the fragment comes from
}
\description{
This method use the cigar and the start mapping position of each reads to generate the position of each spliced alignment
}
\examples{
bamfilein <- system.file("data","s1.chr1.bam",package = "rnaCleanR")
alignment <- GenomicAlignments::readGAlignments(bamfilein) 
alignmentInChr1 <- alignment[seqnames(alignment)=="1"] 
fragments <- getFragment(alignmentInChr1)

}

