---
title: "Using rnaCleanR package"
author: "Thu Hien To & Steve Pederson"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Introduction

rnaCleanR is a package used to quantify and remove double-strand DNA within a strand-specific RNA-seq. It takes as input one or several bam files and use a window slding across the whole genome to compute the amount of positive/negative reads in each window. Then the users can get some plots on those windows: 1) a histogram about the frequency of positive proportion over all windows, which give you an idea on how much double-strand DNA is contained in your sample; 2) a plot on sum of reads vs positive proportion, which give you an idea about which threshold to be used for filtering your sample. Then, the users can filter their RNA sample using an appropriate threshold.

The input bam file must be sorted before and an index file must be given at the same folder. The average read length is required to calculate the size of sliding window and its step. The functions filterOne removes potential double strand DNA from one bam file, while filterMulti filters several bam files. The later method uses the merged alignments from all input bam files to decide which windows are kept; then each individual bam file is filtered based on the same set of kept windows.
 The function getPlot can be used to have an idea of how much the sample was contaminated and which threshold to be used. You can also have these plots at the same time as filtering the bam file by using function filterOne with options histPlot=TRUE or winPlot=TRUE.
 
 It uses a window sliding across each chromosome, and consider the proportion of positive/negative reads in each window comparing to a given threshold, to decide wheather a window is kept or removed. Then each read fragment is assigned to a probability of being kept or not, based on the proportion of positive/negative of the kept windows that contain the fragment.
## Get plots

```{r getplot,  eval=FALSE}
file <- system.file("data","s1.chr1.bam",package = "rnaCleanR")
getPlot(bamfilein = file,readLength = 100,histPlotFile = "hist.pdf",winPlotFile = "win.pdf")
```

Two figures `hist.pdf` and `win.pdf` will be created

## Filter one bam file

```{r filterOne, eval=FALSE}
file <- system.file("data","s1.chr1.bam",package = "rnaCleanR")
filterOne(bamfilein = file,bamfileout = "filter.bam", readLength = 100, threshold = 0.7)
```


## Filter multiple bam files

```{r filterMulti, eval=FALSE}
file <- system.file("data",c("s1.chr1.bam","s2.chr1.bam"),package = "rnaCleanR")
filterMulti(bamfilein = file,bamfileout = c("filter1.bam","filter2.bam"), readLength = 100, threshold = 0.7)

```


