#' @title Plot the histogram of positive proportions of the input windows data frame
#'
#' @description Plot the histogram of positive proportions of the input windows data frame
#'
#' @param windows data frame containing the number of positive/negative reads for each window
#' Windows can be get by the function \code{getWinFromBamFile}.
#'
#' @param group an integer vector that specifies how you want to partition the windows based on the number of reads. By default \code{group} = c(10,100,1000), which means that your windows will be parition into 4 groups, those have number of reads < 10, from 10 to 100, from 100 to 1000, and > 1000
#' @param save if TRUE, then the plot will be save into the file given by \code{file} parameter
#' @param file the file name to save to plot
#' @param facet_wrap_chromosomes if TRUE, then the plots will be splitted by chromosomes. FALSE by default
#' @param useCoverage if TRUE then plot the coverage strand information, otherwise plot the number of reads strand information. FALSE by default
#' @seealso getWinFromBamFile, plotWin
#'
#' @examples
#' \dontrun{
#' #for single end bam file
#' bamfilein = system.file("extdata","s1.chr1.bam",package = "strandCheckR")
#' windows <- getWinFromBamFile(file = bamfilein)
#' plotHist(windows)
#' 
#' #for paired end bamfile
#' bamfilepair = system.file("extdata","120.bam",package = "strandCheckR")
#' windowsP <- getWinFromBamFile(file = bamfilepair)
#' plotHist(windowsP)
#' }
#' 
#' @importFrom magrittr set_colnames
#' @importFrom dplyr mutate select filter
#' @importFrom graphics hist
#' @importFrom stats rbinom
#' @importFrom stats pnorm
#' @export
plotHist <- function(windows,group=c(10,100,1000),save=FALSE,file = "hist.pdf",facet_wrap_chromosomes=FALSE, useCoverage=FALSE){
  if (!facet_wrap_chromosomes){
    windows <- windows[names(windows) != "Chr"]
  }
  if (useCoverage){
    windows <-  windows[!(names(windows) %in% c("NbPositive","NbNegative"))]
    windows <- as.data.frame(windows)
    windows <- mutate(windows,"PositiveProportion" = CovPositive/(CovPositive+CovNegative)) %>%
      select(-c(CovPositive,CovNegative))
  } else{
    windows <- windows[!(names(windows) %in% c("CovPositive","CovNegative"))]
    windows <- as.data.frame(windows)
    windows <- mutate(windows,"PositiveProportion" = NbPositive/(NbPositive+NbNegative)) %>%
      select(-c(NbPositive,NbNegative))
  }
  if (length(group)==0){
    leg <- "all"
  } else{
    leg <- paste0("<",group[1])
    for (i in seq_along(group[-1])){
      leg <- c(leg,paste0(group[i],"-",group[i+1]))
    }
    leg <- c(leg,paste0(">",group[length(group)]))
  }
  x <- lapply(seq_along(group),function(i){which(group[i]<windows$MaxCoverage)})
  G <- rep(leg[1],nrow(windows))
  for (i in seq_along(group)){
    G[x[[i]]] <- leg[i+1]
  }
  windows$MaxCoverage = G
  breaks <- 100
  if ("Type" %in% colnames(windows)){
    histoFirst <- lapply(leg,function(l){
      a <- filter(windows,MaxCoverage==l,Type=="First")
      if (nrow(a)>0){
        if (facet_wrap_chromosomes){
          chromosomes <- unique(a$Chr)
          sapply(chromosomes,function(chr){
            ac <- filter(a,Chr==chr)
            hist(ac$PositiveProportion,breaks=0:(breaks+1)/breaks - 1/(2*breaks),plot=FALSE)$count
          }) %>% reshape2::melt()  %>% set_colnames(c("PositiveProportion","Chr","Count")) %>% mutate("MaxCoverage"=l,"Type"="First")
        } else{
          data.frame("Count"=hist(a$PositiveProportion,breaks=0:(breaks+1)/breaks - 1/(2*breaks),plot=FALSE)$count,
                     "PositiveProportion" = 0:100,
                     "MaxCoverage" = l,"Type"="First")
        }
      }})
    null <- sapply(histoFirst,is.null)
    histoFirst <- histoFirst[!null]
    histoFirst <- do.call(rbind,histoFirst)

    histoSecond <- lapply(leg,function(l){
      a <- filter(windows,MaxCoverage==l,Type=="Second")
      if (nrow(a)>0){
        if (facet_wrap_chromosomes){
          chromosomes <- unique(a$Chr)
          sapply(chromosomes,function(chr){
            ac <- filter(a,Chr==chr)
            hist(ac$PositiveProportion,breaks=0:(breaks+1)/breaks - 1/(2*breaks),plot=FALSE)$count
          }) %>% reshape2::melt()  %>% set_colnames(c("PositiveProportion","Chr","Count")) %>% mutate("MaxCoverage"=l,"Type"="Second")
        } else{
          data.frame("Count"=hist(a$PositiveProportion,breaks=0:(breaks+1)/breaks - 1/(2*breaks),plot=FALSE)$count,
                     "PositiveProportion" = 0:100,
                     "MaxCoverage" = l,"Type"="Second")
        }
      }})
    null <- sapply(histoSecond,is.null)
    histoSecond <- histoSecond[!null]
    histoSecond <- do.call(rbind,histoSecond)
    histo <- rbind(histoFirst,histoSecond)
    if (facet_wrap_chromosomes){
      g <- ggplot2::ggplot(histo, ggplot2::aes(PositiveProportion, Count, fill=MaxCoverage, width=1))+
        ggplot2::scale_fill_discrete(breaks=leg)+
        ggplot2::geom_bar(stat="identity", width=.3,position="stack") +
        ggplot2::facet_wrap(~Type+Chr)
    }
    else{
      g <- ggplot2::ggplot(histo, ggplot2::aes(PositiveProportion, Count, fill=MaxCoverage, width=1))+
        ggplot2::scale_fill_discrete(breaks=leg)+
        ggplot2::geom_bar(stat="identity", width=.3,position="stack")+
        ggplot2::facet_wrap(~Type)
    }
     g <- g +  ggplot2::ylab("Count") +
      ggplot2::xlab("Positive Proportion")+
      ggplot2::theme_bw()
    if (useCoverage){
      g <- g + ggplot2::labs(fill = "Max Coverage")
    }
    if (save==TRUE){
      message("The plot will be saved to the file ",file)
      ggplot2::ggsave(filename = file)
    }
    else{
     g
    }
  }
  else{
    histo<- lapply(leg,function(l){
      a <- filter(windows,MaxCoverage==l) 
      if (nrow(a)>0){
        if (facet_wrap_chromosomes){
          chromosomes <- unique(a$Chr)
          sapply(chromosomes,function(chr){
            ac <- filter(a,Chr==chr)
            hist(ac$PositiveProportion,breaks=0:(breaks+1)/breaks - 1/(2*breaks),plot=FALSE)$count
          }) %>% reshape2::melt()  %>% set_colnames(c("PositiveProportion","Chr","Count")) %>% mutate("MaxCoverage"=l)
        }
        else{
          data.frame("Count"=hist(a$PositiveProportion,breaks=0:(breaks+1)/breaks - 1/(2*breaks),plot=FALSE)$count,
                     "PositiveProportion" = 0:100,
                     "MaxCoverage" = l)
        }
      }})
    rm(windows)
    null <- sapply(histo,is.null)
    histo <- histo[!null]
    histo <- do.call(rbind,histo)
    if (facet_wrap_chromosomes){
      g <- ggplot2::ggplot(histo, ggplot2::aes(PositiveProportion, Count, fill=MaxCoverage, width=1))+
        ggplot2::scale_fill_discrete(breaks=leg)+
        ggplot2::facet_wrap(~Chr)
    }
    else{
      g <- ggplot2::ggplot(histo, ggplot2::aes(PositiveProportion, Count, fill=MaxCoverage, width=1))+
        ggplot2::scale_fill_discrete(breaks=leg)
    }
    g <- g + ggplot2::geom_bar(stat="identity", width=.3,position="stack") +
      ggplot2::ylab("Count") +
      ggplot2::xlab("Positive Proportion")+
      ggplot2::theme_bw()
    if (useCoverage){
      g <- g + ggplot2::labs(fill = "Max Coverage")
    }
    if (save==TRUE){
      message("The plot will be saved to the file ",file)
       ggplot2::ggsave(filename = file,plot = g)
    }
    else{
      g
    }
  }
}
