# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' @title  Compute strand information of sliding window (plot version)
#'
#' @description Compute the positive proportion, sum of reads, max coverage and the group of each window. Windows are grouped based on their maximum coverage: by default definition, groups spead from 1 to 4, which correspond to the max coverage respectively in the range "0-10","10-100","100-1000",">1000"
#' This method is used in the getPlot function when we only need the information to plot, and do not need to filter the reads afterward.
#'
#' @param covLen the run length of an Rle object which is the coverage comes from positive reads
#' @param covVal the run value of an Rle object which is the coverage comes from positive reads
#' @param covNegLen the run length of an Rle object which is the coverage comes from negative reads
#' @param covNegVal the run value of an Rle object which is the coverage comes from negative reads
#' @param end the last base on the reference chromosome that the sliding window atteints
#' @param readLength the average length of reads
#' @param win the size of the sliding window
#' @param step the step of the sliding window
#' @param minCov if a window has the max coverage least than minCov, then it will not be counted
#'
#' @return A data frame which contains the information of all windows: Starting positive, Number of Positive/Negative Reads, the Maximum Coverage.
#' @examples
#' bamfilein <- system.file("data","s1.chr1.bam",package = "rnaCleanR")
#' alignment <- GenomicAlignments::readGAlignments(bamfilein)
#' alignmentInChr1 <- alignment[seqnames(alignment)=="1"]
#' covPos <- alignmentInChr1[strand(alignment)=="+"] %>% GenomicAlignments::coverage()
#' covNeg <- alignmentInChr1[strand(alignment)=="-"] %>% GenomicAlignments::coverage()
#' len <- length(covChr)
#' readLength <- 100
#' win <- 1000
#' step <- 100
#' minCov <- 0
#' windows <- rnaCleanR::computeWinInfo(runLength(covPos),runValue(covPos),runLength(covNeg),runValue(covNeg),readLength,len,win,step,minCov)
#'
#'
#' @export
#'
computeWinInfo <- function(covLen, covVal, end, win, step) {
    .Call('rnaCleanR_computeWinInfo', PACKAGE = 'rnaCleanR', covLen, covVal, end, win, step)
}

